stacker/actions/base.py

```diff
@@ -154,19 +120,107 @@ def __init__(self, context, provider_builder=None, cancel=None):
         self.bucket_region = context.config.stacker_bucket_region
         if not self.bucket_region and provider_builder:
             self.bucket_region = provider_builder.region
-        self.s3_conn = get_session(self.bucket_region).client('s3')
+        self.s3_conn = (getattr(self.context, 's3_conn', None) or
```

Comment:
I have an S3 client being created in Context to be able to manage/check the persistent graph through execution. This is to reuse that client or create a new one if it does not exist on Context.

iirc there are some tests that break if I had done `self.s3_conn = context.s3_conn` so I went this route rather than changing the tests/mock Context objects.